runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,449 @@
+     * @param tableRegistry Table registry from the corfu client.
+     * @param objectsView   Objects View from the corfu client.
+     * @param namespace     Namespace to perform the queries within.
+     */
+    public Query(final TableRegistry tableRegistry, final ObjectsView objectsView, final String namespace) {
+        this.tableRegistry = tableRegistry;
+        this.objectsView = objectsView;
+        this.namespace = namespace;
+    }
+
+    private Token getToken(Timestamp timestamp) {
+        return new Token(timestamp.getEpoch(), timestamp.getSequence());
+    }
+
+    private void txBegin(Timestamp timestamp) {
+        TransactionBuilder transactionBuilder = objectsView.TXBuild()
+                .type(TransactionType.SNAPSHOT);
+
+        if (timestamp != null) {
+            transactionBuilder.snapshot(getToken(timestamp));
+        }
+
+        transactionBuilder.build().begin();
+    }
+
+    private void txEnd() {
+        objectsView.TXEnd();
+    }
+
+    private <K extends Message, V extends Message, M extends Message>
+    Table<K, V, M> getTable(@Nonnull final String tableName) {
+        return tableRegistry.getTable(this.namespace, tableName);
+    }
+
+    /**
+     * Fetch the Value payload for the specified key.
+     *
+     * @param tableName Table name.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of value.
+     * @return Value.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> get(@Nonnull final String tableName,
+                          @Nonnull final K key) {
+        return get(tableName, null, key);
+    }
+
+    /**
+     * Fetch the Value payload for the specified key at the specified snapshot.
+     *
+     * @param tableName Table name.
+     * @param timestamp Timestamp to perform the query.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of value.
+     * @return Value.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> get(@Nonnull final String tableName,
+                          @Nullable final Timestamp timestamp,
+                          @Nonnull final K key) {
+        try {
+            txBegin(timestamp);
+            return ((Table<K, V, M>) getTable(tableName)).get(key);
+        } finally {
+            txEnd();
+        }
+    }
+
+    /**
+     * Fetch the Value payload with the metadata for the specified key.
+     *
+     * @param tableName Table name.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of value.
+     * @return Corfu Record.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> getRecord(@Nonnull final String tableName,
+                                @Nonnull final K key) {
+        return get(tableName, null, key);
```

Comment:
Should this be `return getRecord(tableName, null, key)`?

Reply:
yes, thanks!